odoo/tools/misc.py

```diff
@@ -1821,6 +1821,44 @@ def verify_hash_signed(env, scope, payload):
     return None
 
 
+def binary_token(record, field_name, timestamp=None):
+    """Generate a token granting access for at least 14 days from the binary
+    routes (/web/content or /web/image) to the given record and field_name.
+    The token is deterministic within a 14-day period (even across different
+    days/months/years) to allow browser caching, and expires after maximum 28
+    days to prevent infinite access. Different record/field combinations will
+    expire at different times to prevent thundering herd problems."""
+    record.ensure_one()
+    if not timestamp:
+        now = datetime.datetime.now()
+        epoch = datetime.datetime(1970, 1, 1)
+        start_of_period = epoch + datetime.timedelta(days=((now - epoch).days // 14) * 14)
+        unique_str = f"{record._name}-{str(record.id)}-{field_name}"
+        days_offset = int(hashlib.sha256(f"d-{unique_str}".encode()).hexdigest(), 16)
+        seconds_offset = int(hashlib.sha256(f"s-{unique_str}".encode()).hexdigest(), 16)
+        date = start_of_period + datetime.timedelta(
+            days=days_offset % 14, seconds=seconds_offset % (24 * 60 * 60)
+        )
+        while date - datetime.timedelta(days=14) < now:
+            date += datetime.timedelta(days=14)
```

Comment:
The idea was to cap at 28, and the loop was only called twice maximum, so it could have been one +14 (always) and one condition +14.

If we're ok with the extra range (14-42 days) rather than (14-28 days) then we can always do +28 (rather than +14 and conditionally +14 again).